drivers/gpio/gpio_eos_s3.c

```diff
@@ -0,0 +1,325 @@
+}
+
+static int gpio_eos_s3_port_set_masked_raw(const struct device *dev,
+					   uint32_t mask,
+					   uint32_t value)
+{
+	ARG_UNUSED(dev);
+	uint32_t target_value;
+	uint32_t output_states = MISC_CTRL->IO_OUTPUT;
+
+	target_value = ((output_states & ~mask) | (value & mask));
+	MISC_CTRL->IO_OUTPUT = (target_value & GPIOS_MASK);
+
+	return 0;
+}
+
+static int gpio_eos_s3_port_set_bits_raw(const struct device *dev,
+					 uint32_t mask)
+{
+	ARG_UNUSED(dev);
+
+	MISC_CTRL->IO_OUTPUT |= (mask & GPIOS_MASK);
+
+	return 0;
+}
+
+static int gpio_eos_s3_port_clear_bits_raw(const struct device *dev,
+					   uint32_t mask)
+{
+	ARG_UNUSED(dev);
+
+	MISC_CTRL->IO_OUTPUT &= ~(mask & GPIOS_MASK);
+
+	return 0;
+}
+
+static int gpio_eos_s3_port_toggle_bits(const struct device *dev,
+					uint32_t mask)
+{
+	ARG_UNUSED(dev);
+	uint32_t target_value;
+	uint32_t output_states = MISC_CTRL->IO_OUTPUT;
+
+	target_value = output_states ^ mask;
+	MISC_CTRL->IO_OUTPUT = (target_value & GPIOS_MASK);
+
+	return 0;
+}
+
+static int gpio_eos_s3_manage_callback(const struct device *dev,
+				       struct gpio_callback *callback, bool set)
+{
+	struct gpio_eos_s3_data *data = dev->data;
+
+	return gpio_manage_callback(&data->callbacks, callback, set);
+}
+
+static int gpio_eos_s3_pin_interrupt_configure(const struct device *dev,
+					       gpio_pin_t gpio_num,
+					       enum gpio_int_mode mode,
+					       enum gpio_int_trig trig)
+{
+	struct gpio_eos_s3_data *data = dev->data;
+	GPIOCfgTypeDef gpio_cfg;
+	PadConfig pad_config = gpio_eos_s3_pad_select(dev, gpio_num);
+
+	gpio_cfg.ucGpioNum = gpio_num;
+	gpio_cfg.xPadConf = &pad_config;
+
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		/* Reset IRQ configuration */
+		int irq_num = HAL_GPIO_IntrCfg(&gpio_cfg);
+
+		if (irq_num < 0) {
+			return -EINVAL;
+		}
+
+		data->gpio_irq_num[gpio_num] = BIT((uint8_t)irq_num);
+		/* Disable IRQ */
+		INTR_CTRL->GPIO_INTR_EN_M4 &= ~data->gpio_irq_num[gpio_num];
+	} else {
+		/* Prepare configuration */
+		if (mode == GPIO_INT_MODE_LEVEL) {
+			gpio_cfg.intr_type = LEVEL_TRIGGERED;
+			if (trig == GPIO_INT_TRIG_LOW) {
+				gpio_cfg.pol_type = FALL_LOW;
+			} else {
+				gpio_cfg.pol_type = RISE_HIGH;
+			}
+		} else {
+			gpio_cfg.intr_type = EDGE_TRIGGERED;
+			switch (trig) {
+			case GPIO_INT_TRIG_LOW:
+				gpio_cfg.pol_type = FALL_LOW;
+				break;
+			case GPIO_INT_TRIG_HIGH:
+				gpio_cfg.pol_type = RISE_HIGH;
+				break;
+			case GPIO_INT_TRIG_BOTH:
+				return -ENOTSUP;
+			}
+		}
+
+		/* Set IRQ configuration */
+		int irq_num = HAL_GPIO_IntrCfg(&gpio_cfg);
+
+		if (irq_num < 0) {
+			return -EINVAL;
+		}
+
+		data->gpio_irq_num[gpio_num] = BIT((uint8_t)irq_num);
+
+		/* Enable IRQ */
+		INTR_CTRL->GPIO_INTR_EN_M4 |= data->gpio_irq_num[gpio_num];
+	}
+
+	return 0;
+}
+
+static void gpio_eos_s3_isr(const struct device *dev)
+{
+	struct gpio_eos_s3_data *data = dev->data;
+	uint8_t intr_status = INTR_CTRL->GPIO_INTR & GPIOS_MASK;
+	
+	if (!intr_status) {
+		return;
+	}
```

Comment:
I'm not familiar with the architecture and I may be wrong but that doesn't look right. If you entered the `gpio_eos_s3_isr` it means there was some interrupt that was active. It's not an interrupt that we find interesting since it's outside of `GPIOS_MASK`, however it is there. You didn't clear it. Soon after you exit the interrupt routine you'll enter it again, the interrupt is still marked as active. In principle your program will freeze here in the endless loop.

On architectures that do not automatically clear interrupt status register on read operation - and it looks like this is the one - it's best to do it manually just after you read the status register. It also ensures that if there are two interrupt events close to each other and the interrupt routine is long both interrupts will be noticed. If you clear status register late you may miss an interrupt.